python-checks/src/main/java/org/sonar/python/checks/ModifiedParameterValueCheck.java

```diff
@@ -0,0 +1,264 @@
+    MUTATING_METHODS.put("collections.UserDict", DICT_MUTATING_METHODS);
+    MUTATING_METHODS.put("collections.ChainMap", DICT_MUTATING_METHODS);
+    MUTATING_METHODS.put("collections.Counter", COUNTER_MUTATING_METHODS);
+    MUTATING_METHODS.put("collections.OrderedDict", ORDERED_DICT_MUTATING_METHODS);
+    MUTATING_METHODS.put("collections.defaultdict", DEFAULT_DICT_MUTATING_METHODS);
+  }
+
+  @Override
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(FUNCDEF, ctx -> {
+      FunctionDef functionDef = (FunctionDef) ctx.syntaxNode();
+      // avoid raising issues on nested function, it may have been done on purpose
+      if (TreeUtils.firstAncestorOfKind(functionDef, FUNCDEF) != null) {
+        return;
+      }
+      for (Parameter parameter : nonTupleParameters(functionDef)) {
+
+        List<Tree> mutations = getMutations(parameter);
+        if (!mutations.isEmpty()) {
+          PreciseIssue preciseIssue = ctx.addIssue(parameter, "Change this default value to \"None\" and initialize this parameter inside the function/method.");
+          mutations.forEach(mutation -> preciseIssue.secondary(mutation, null));
+        }
+      }
+    });
+  }
+
+  @CheckForNull
+  private static String defaultValueType(Expression expression) {
+    for (String nonCompliantType : MUTATING_METHODS.keySet()) {
+      if (expression.type().canOnlyBe(nonCompliantType)) {
+        return nonCompliantType;
+      }
+    }
+    return null;
+  }
+
+  private static boolean isUsingMemoization(Symbol symbol) {
+    return symbol.name().contains("cache") || symbol.name().contains("memo");
+  }
+
+  private static List<Tree> getMutations(Parameter parameter) {
+    Expression defaultValue = parameter.defaultValue();
+    if (defaultValue == null) {
+      return Collections.emptyList();
+    }
+
+    Optional<Symbol> paramSymbol = getSymbolFromTree(parameter.name());
+    if (!paramSymbol.isPresent() || isUsingMemoization(paramSymbol.get())) {
+      return Collections.emptyList();
+    }
+
+    if (!defaultValue.type().canOnlyBe(BuiltinTypes.NONE_TYPE)) {
+      List<Tree> attributeSet = getAttributeSet(paramSymbol.get());
+      if (!attributeSet.isEmpty()) {
+        return attributeSet;
+      }
+    }
+    String defaultValueType = defaultValueType(defaultValue);
+    Set<String> typeMutatingMethods = MUTATING_METHODS.get(defaultValueType);
+    if (typeMutatingMethods == null) {
+      return Collections.emptyList();
+    }
+    return paramSymbol.get().usages().stream().filter(usage -> {
+      Tree parent = usage.tree().parent();
+      if (parent.is(QUALIFIED_EXPR)) {
+        QualifiedExpression qualifiedExpression = (QualifiedExpression) parent;
+        return  getSymbolFromTree(qualifiedExpression.qualifier()).equals(paramSymbol) && isMutatingMethod(typeMutatingMethods, qualifiedExpression.name().name());
+      }
+      return isUsedInDelStatement(usage.tree()) ||
```

Comment:
I find this block of code hard to read due to all the returns and tight formatting. I think it should be a little bit more spacious, I would introduce some blank lines between the control structures.

Reply:
I decided to extract a function to make it more readable. Thanks!